src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,294 @@
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier?.EscapeIdentifier();
+
+        protected override async Task<(bool, NamespaceDeclarationSyntax)> ShouldPositionTriggerRefactoringAsync(
+            Document document, 
+            int position, 
+            CancellationToken cancellationToken)
+        {
+            var compilationUnit = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false) as CompilationUnitSyntax;            
+            var namespaceDecls = compilationUnit.DescendantNodes().OfType<NamespaceDeclarationSyntax>().ToImmutableArray();
+
+            // Here's conditions that trigger the refactoring (all have to be true in each scenario):
+            // 
+            // - There's only one namespace declaration in the document and all types are declared in it:
+            //    1. No nested namespace declaration (even it's empty).
+            //    2. The cursor is on the name of the namespace declaration.
+            //    3. The name of the namespace is valid (i.e. no errors).
+            //    4. No partial type declared in the namespace. Otherwise its multiple declaration will
+            //       end up in different namespace.
+            //
+            // - There's no namespace declaration and all types in the document are declared in global namespace:
+            //    1. The cursor is on the name of first declared type.
+            //    2. No partial type declared in the document. Otherwise its multiple declaration will
+            //       end up in different namespace.
+
+            if (namespaceDecls.Length == 1 && compilationUnit.Members.Count == 1)
+            {
+                var namespaceDeclaration = namespaceDecls.Single();
+                Debug.Assert((object)namespaceDeclaration == (object)compilationUnit.Members.Single());
+
+                var shouldTrigger = namespaceDeclaration.Name.Span.IntersectsWith(position) 
+                    && namespaceDeclaration.Name.GetDiagnostics().All(diag => diag.DefaultSeverity != DiagnosticSeverity.Error);
+
+                shouldTrigger = shouldTrigger && 
+                    !(await ContainsPartialTypeWithMultipleDeclarationsAsync(document, namespaceDeclaration, cancellationToken)
+                    .ConfigureAwait(false));
+
+                return (shouldTrigger, shouldTrigger ? namespaceDeclaration : null);
+            }
+
+            if (namespaceDecls.Length == 0)
+            {
+                var firstMemberDeclaration = compilationUnit.Members.FirstOrDefault();
+
+                var shouldTrigger = firstMemberDeclaration != null
+                    && firstMemberDeclaration.GetNameToken().Span.IntersectsWith(position);
+
+                shouldTrigger = shouldTrigger &&
+                    !(await ContainsPartialTypeWithMultipleDeclarationsAsync(document, compilationUnit, cancellationToken)
+                    .ConfigureAwait(false));
```

Comment:
fwiw, i think a simpelr way to do this would be to just extract a method that return a Namespcae or CompilationUnit if things are legal.  This method will then call that method.  This method will check if it gets anything back, if not, it fails.  If so, it checks the namespace/compilation unit for types with multiple declarations.  And finally, in the compilation unit case, it returns null back.

--

Alternatively, don't even return null back in the success case.  Just have this method return the namespacedecl/compilation unit if it succeeds.  The caller of that topmost method can decide what to do (i.e. gets nothing back, it bails out.  gets namespace decl back, uses it.  gets compilation unit back, doesn't use it).

Reply:
makes sense.